### PR TITLE
Finish making `InputControl` et al. more controllable

### PR DESCRIPTION
**Pull request by stokesman**
## What?
Builds upon #40518 to add the changes to `RangeControl` ~~and `UnitControl`~~ necessitated by the making `InputControl` controllable regardless of its focused state.

## Why?
`RangeControl` needs changes is because it depends on its `NumberControl` being self-controlled while focused (i.e. ignoring the `value` prop it gets from `RangeControl`) so that when `RangeControl` actively constrains its state value it does not interfere with the entry of the value in the focused input.

~~I've not yet pinpointed why `UnitControl` requires a change but without it, the issue is that it calls `onChange` twice for a single change (from a `blur` event). That probably wouldn't have hurt a thing but it's nice that our unit tests caught this.~~ UPDATE: It turns out this was an existing issue but our test that would catch it, wasn't doing so on trunk. The related changes from this PR have already been extracted/merged with #40796 so they are to be removed from this branch.

## How?
- Logic is added to `RangeControl` to temporarily keep entered values in its number input if they are out-of-range or non-numeric.
- ~~`UnitControl` stops calling `onChange` explicitly from `mayUpdateUnit` because apparently `InputControl` now does so on its own~~

## Testing Instructions
Unit tests pass.
Try finding any differences manually testing `RangeControl` ~~and `UnitControl`~~.

### CODE REVIEW

**Comment by ciampo** [comment]
I just pushed a [quick fix](https://github.com/WordPress/gutenberg/pull/40568/commits/295035c67e25f4f2b460bb22804a93b1d2b00bd9) to make sure that the CI checks could complete and [added a small comment](https://github.com/WordPress/gutenberg/pull/40568/commits/739372f1e5f721e98530f2ef2007724241e24dbe) to explain why the `onChange` callback isn't called explicitly in `UnitControl`'s `mayUpdateUnit` function

**Review by ciampo — COMMENTED** [review]
Thank you @stokesman for you work on this!

Overall the code changes all make sense to me, and so my idea would be to merge this ASAP into #40568 (cc @youknowriad ) so that we can all give a final look at the fix on that PR.

The good amount of unit tests that we have also reassures me that these changes should be working fine.

I spent some more time checking the components in Storybook, and noticed a regression in the behavior of the "Reset" button in `RangeControl`:

`trunk`:

https://user-images.githubusercontent.com/1083581/165085172-2999ab61-95ac-438d-aa68-8dfa462a5305.mp4

This PR:

https://user-images.githubusercontent.com/1083581/165085165-094b6ca2-f00f-4c49-82d2-0639a6311733.mp4


Basically, with the changes from this PR, the "Reset" button needs to be clicked twice before effectively resetting the value. Once this is ironed out, it would be good to add a unit test to `RangeControl` to avoid this regression in the future.

**Comment by stokesman** [comment]
Thanks for the fixups and testing @ciampo!

>the "Reset" button needs to be clicked twice before effectively resetting the value. Once this is ironed out, it would be good to add a unit test to RangeControl to avoid this regression in the future.

Great catch. It also made me realize the (visual) initial position was broken for the `RangeControl`’s without values. Besides that, this turned into more ironing than I would have guessed. Here's how it played out.

cd1b4df
We have a unit test that should catch the reset regression. I found that it fails only if the component is controlled (like Storybook). So this commit updates the test to run for both controlled and uncontrolled instances.

9a3804f
In looking to fix the reset regression, I noticed the `handleOnReset` function was calling both `setValue` and `onChange` but the controlled value hook is meant to handle calling `onChange`. I updated it to do so and this fixed the issue in Storybook yet the unit test continued to fail 😕.

4dc4674 & 302096d
I noticed the logical OR operator in the reset handling logic of `InputControl`’s reducer seemed likely to be part of the issue. I tried changing it so falsy values could be used in the reset and it that got the unit test passing. However, it surfaced or worsened an issue with the state sync logic.

f3c881f
Corrects the initial position visuals of `RangeControl`.

19fd100 & 68a02ea
To keep the two effect hooks used to sync state from getting into fights the event is used as the primary condition for which hook runs. This can work because all actions have events associated and an update incoming from props does not need one. The current implementation does pass an event through on such updates but it's the leftover from the last action and serves no apparent purpose. This also removes the event from the state because that too seems like something that was just so and not really of use or in use. This all might have been done a little differently but I tried to do it in as few changes as possible. This may have also obviated 4dc4674/ 302096d but that seems like something to keep anyway.

d6b43bb
Just some cleanup that can be done since React 17.

**Comment by stokesman** [comment]
Oh, and this is still not there yet. I've noticed an issue with syncs after drag actions but 🤞 hopefully that's something simple I've missed. Calling it a day for now.

**Comment by aaronrobertshaw** [comment]
> Oh, and this is still not there yet. I've noticed an issue with syncs after drag actions but 🤞 hopefully that's something simple I've missed. Calling it a day for now.

I'd just encountered the issue after drag actions and was recording a demo video of it. I'm out of time for today as well but can help take a look tomorrow if you don't get to it.


https://user-images.githubusercontent.com/60436221/165251719-8ea9a124-55f8-43e7-90af-b99eda68bd0d.mp4

File: packages/components/src/input-control/reducer/reducer.ts [hunk 1]
```diff
@@ -54,6 +54,15 @@ function inputControlStateReducer(
 	return ( state, action ) => {
 		const nextState = { ...state };
 
+		// Updates state and returns early when there's no action type. These
+		// are controlled updates and need no exposure to additional reducers.
+		if ( ! ( 'type' in action ) ) {
+			let { value = state.value } = action;
+			value ??= '';
+			if ( value !== '' ) value = `${ value }`;
+			return { ...state, value };
```
Comment (mirka) [inline packages/components/src/input-control/reducer/reducer.ts]: Nit: As a code reader I would find a simpler version like this to be easier to verify the logic 👇

```suggestion
			return {
				...state,
				value: `${ action.value ?? '' }`,
			};
```

(Or something like that, anything that removes the let/if)

**Review by mirka — COMMENTED** [review]
Great work here as usual, changes sound reasonable to me too and the test coverage is reassuring. Once the remaining reset issue is addressed, let's get this merged into Riad's branch like Marco suggests!

**Comment by stokesman** [comment]
Thank you @mirka  and @aaronrobertshaw for having a look. 

>issue with syncs after drag actions but 🤞 hopefully that's something simple I've missed

Fixed in 547a7c7. It was pretty simple. The `dragEnd` action doesn't change the `state.value` so the propagating effect hook wouldn't run, the event wouldn't clear and the intaking effect hook would skip. Adding `state.isDragging` to the propagating effect hook’s dependencies was the solution. The clearing of the event was also made more sure by keeping it independent of the other conditions for calling `onChange`.

**Review by aaronrobertshaw — APPROVED** [review]
LGTM 👍 

The [issue with drag actions](https://github.com/WordPress/gutenberg/pull/40568#issuecomment-1109478912) has been resolved and the various components are working well for me.

https://user-images.githubusercontent.com/60436221/165425661-312e0aa6-d7a1-4037-98f6-0c0b033ea8dc.mp4

**Comment by ciampo** [comment]
Thanks for the amazing work, @stokesman (and for the very detailed explanation of the issues you encountered and solved)

I had another attempt at testing `RangeControl`'s behavior, and discovered another little issue around the resetting behavior:

On `trunk`, resetting the component sets the `value` to `initialPosition` (if set):

https://user-images.githubusercontent.com/1083581/165486845-404a8d4a-64f0-48c4-b1b8-d179396edd06.mp4

On this PR, hitting the reset button aways clears the value, ignoring the `initialPosition`:

https://user-images.githubusercontent.com/1083581/165487222-b4c2c93a-9c6b-4046-bb39-625b5cf1db10.mp4

---

It would be great if we could also add another quick **unit test** to avoid regressions on this front too.

I'm sorry if this PR is taking longer that expected, but the logic around these components is quite complicated and therefore it's easy to introduce unwanted behaviors when making changes.

**Comment by stokesman** [comment]
Thanks for the encouraging words and the diligent testing @ciampo.

>I had another attempt at testing RangeControl's behavior, and discovered another little issue around the resetting behavior…It would be great if we could also add another quick unit test to avoid regressions on this front too.

I'll add that test and raise you one more 06d16fd. We had a test for reset to `resetFallbackValue`. This adds one for `initialPosition` and one for if both are defined. It definitely seems a bit absurd that we have both of those props. The big difference is that when reset with only `initialPosition` defined the value passed to `onChange` is `undefined` (not the value of `initialPosition`).

Besides the reset issue I've noticed there is potential interference with value entry in the number input.

##### In Trunk with `min: 10, max: 200`
https://user-images.githubusercontent.com/9000376/165815161-5f289b9a-d3cb-469f-b61a-f18d54523a0f.mp4

##### In this branch with `min: 10, max: 200`
https://user-images.githubusercontent.com/9000376/165815551-78071481-c0a7-4f10-9240-93d31386231d.mp4

We have a test that is supposed to catch this and it proves to be yet another instance of it only failing for controlled instances. To avoid anymore of that in other tests, ce0ac42 makes all the tests run in both modes.

The test failures can be verified in this branch for now.

<details><summary>A patch applicable to trunk with the same changes to range-control’s unit tests to verify that they all pass.</summary>

```diff
diff --git a/packages/components/src/range-control/test/index.js b/packages/components/src/range-control/test/index.js
index c60f12b3b7..634db9bc8f 100644
--- a/packages/components/src/range-control/test/index.js
+++ b/packages/components/src/range-control/test/index.js
@@ -3,6 +3,11 @@
  */
 import { fireEvent, render } from '@testing-library/react';
 
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
@@ -15,14 +20,26 @@ const getNumberInput = ( container ) =>
 const getResetButton = ( container ) =>
 	container.querySelector( '.components-range-control__reset' );
 
-describe( 'RangeControl', () => {
+function ControlledRangeControl( props ) {
+	const [ value, setValue ] = useState( props.value );
+	const onChange = ( v ) => {
+		setValue( v );
+		props.onChange?.( v );
+	};
+	return <RangeControl { ...props } onChange={ onChange } value={ value } />;
+}
+
+describe.each( [
+	[ 'uncontrolled', RangeControl ],
+	[ 'controlled', ControlledRangeControl ],
+] )( 'RangeControl %s', ( ...modeAndComponent ) => {
+	const [ , Component ] = modeAndComponent;
+
 	describe( '#render()', () => {
 		it( 'should trigger change callback with numeric value', () => {
 			const onChange = jest.fn();
 
-			const { container } = render(
-				<RangeControl onChange={ onChange } />
-			);
+			const { container } = render( <Component onChange={ onChange } /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -39,10 +56,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should render with icons', () => {
 			const { container } = render(
-				<RangeControl
-					beforeIcon="format-image"
-					afterIcon="format-video"
-				/>
+				<Component beforeIcon="format-image" afterIcon="format-video" />
 			);
 
 			const beforeIcon = container.querySelector(
@@ -59,7 +73,7 @@ describe( 'RangeControl', () => {
 
 	describe( 'validation', () => {
 		it( 'should not apply if new value is lower than minimum', () => {
-			const { container } = render( <RangeControl min={ 11 } /> );
+			const { container } = render( <Component min={ 11 } /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -71,7 +85,7 @@ describe( 'RangeControl', () => {
 		} );
 
 		it( 'should not apply if new value is greater than maximum', () => {
-			const { container } = render( <RangeControl max={ 20 } /> );
+			const { container } = render( <Component max={ 20 } /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -85,7 +99,7 @@ describe( 'RangeControl', () => {
 		it( 'should not call onChange if new value is invalid', () => {
 			const onChange = jest.fn();
 			const { container } = render(
-				<RangeControl onChange={ onChange } min={ 10 } max={ 20 } />
+				<Component onChange={ onChange } min={ 10 } max={ 20 } />
 			);
 
 			const numberInput = getNumberInput( container );
@@ -99,7 +113,7 @@ describe( 'RangeControl', () => {
 		it( 'should keep invalid values in number input until loss of focus', () => {
 			const onChange = jest.fn();
 			const { container } = render(
-				<RangeControl onChange={ onChange } min={ -1 } max={ 1 } />
+				<Component onChange={ onChange } min={ -1 } max={ 1 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -118,7 +132,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should validate when provided a max or min of zero', () => {
 			const { container } = render(
-				<RangeControl min={ -100 } max={ 0 } />
+				<Component min={ -100 } max={ 0 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -133,7 +147,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should validate when min and max are negative', () => {
 			const { container } = render(
-				<RangeControl min={ -100 } max={ -50 } />
+				<Component min={ -100 } max={ -50 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -154,11 +168,7 @@ describe( 'RangeControl', () => {
 		it( 'should take into account the step starting from min', () => {
 			const onChange = jest.fn();
 			const { container } = render(
-				<RangeControl
-					onChange={ onChange }
-					min={ 0.1 }
-					step={ 0.125 }
-				/>
+				<Component onChange={ onChange } min={ 0.1 } step={ 0.125 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -179,9 +189,7 @@ describe( 'RangeControl', () => {
 
 	describe( 'initialPosition / value', () => {
 		it( 'should render initial rendered value of 50% of min/max, if no initialPosition or value is defined', () => {
-			const { container } = render(
-				<RangeControl min={ 0 } max={ 10 } />
-			);
+			const { container } = render( <Component min={ 0 } max={ 10 } /> );
 
 			const rangeInput = getRangeInput( container );
 
@@ -190,7 +198,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should render initialPosition if no value is provided', () => {
 			const { container } = render(
-				<RangeControl initialPosition={ 50 } />
+				<Component initialPosition={ 50 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -200,7 +208,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should render value instead of initialPosition is provided', () => {
 			const { container } = render(
-				<RangeControl initialPosition={ 50 } value={ 10 } />
+				<Component initialPosition={ 50 } value={ 10 } />
 			);
 
 			const rangeInput = getRangeInput( container );
@@ -211,7 +219,7 @@ describe( 'RangeControl', () => {
 
 	describe( 'input field', () => {
 		it( 'should render an input field by default', () => {
-			const { container } = render( <RangeControl /> );
+			const { container } = render( <Component /> );
 
 			const numberInput = getNumberInput( container );
 
@@ -220,7 +228,7 @@ describe( 'RangeControl', () => {
 
 		it( 'should not render an input field, if disabled', () => {
 			const { container } = render(
-				<RangeControl withInputField={ false } />
+				<Component withInputField={ false } />
 			);
 
 			const numberInput = getNumberInput( container );
@@ -229,7 +237,7 @@ describe( 'RangeControl', () => {
 		} );
 
 		it( 'should render a zero value into input range and field', () => {
-			const { container } = render( <RangeControl value={ 0 } /> );
+			const { container } = render( <Component value={ 0 } /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -239,7 +247,7 @@ describe( 'RangeControl', () => {
 		} );
 
 		it( 'should update both field and range on change', () => {
-			const { container } = render( <RangeControl /> );
+			const { container } = render( <Component /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -258,7 +266,7 @@ describe( 'RangeControl', () => {
 		} );
 
 		it( 'should reset input values if next value is removed', () => {
-			const { container } = render( <RangeControl /> );
+			const { container } = render( <Component /> );
 
 			const rangeInput = getRangeInput( container );
 			const numberInput = getNumberInput( container );
@@ -274,14 +282,30 @@ describe( 'RangeControl', () => {
 	} );
 
 	describe( 'reset', () => {
-		it( 'should reset to a custom fallback value, defined by a parent component', () => {
+		it.concurrent.each( [
+			[
+				'initialPosition if it is defined',
+				{ initialPosition: 21 },
+				[ '21', undefined ],
+			],
+			[
+				'resetFallbackValue if it is defined',
+				{ resetFallbackValue: 34 },
+				[ '34', 34 ],
+			],
+			[
+				'resetFallbackValue if both it and initialPosition are defined',
+				{ initialPosition: 21, resetFallbackValue: 34 },
+				[ '34', 34 ],
+			],
+		] )( 'should reset to %s', ( ...all ) => {
+			const [ , propsForReset, [ expectedValue, expectedChange ] ] = all;
 			const spy = jest.fn();
 			const { container } = render(
-				<RangeControl
-					initialPosition={ 10 }
+				<Component
 					allowReset={ true }
 					onChange={ spy }
-					resetFallbackValue={ 33 }
+					{ ...propsForReset }
 				/>
 			);
 
@@ -291,14 +315,14 @@ describe( 'RangeControl', () => {
 
 			fireEvent.click( resetButton );
 
-			expect( rangeInput.value ).toBe( '33' );
-			expect( numberInput.value ).toBe( '33' );
-			expect( spy ).toHaveBeenCalledWith( 33 );
+			expect( rangeInput.value ).toBe( expectedValue );
+			expect( numberInput.value ).toBe( expectedValue );
+			expect( spy ).toHaveBeenCalledWith( expectedChange );
 		} );
 
 		it( 'should reset to a 50% of min/max value, of no initialPosition or value is defined', () => {
 			const { container } = render(
-				<RangeControl
+				<Component
 					initialPosition={ undefined }
 					min={ 0 }
 					max={ 100 }

```
</details>

I've also experimented some with fixing these issues and have understood their causes but haven't yet found the minimal solution.

**Comment by stokesman** [comment]
So the remaining two `RangeControl` issues were resets to `initialPosition` and potential interference with value entry in the number input. I think everything is finally solved here. Here's bit more info on the latest commits.

3800d28, ff6e10a & e427a0e
These relate to fixing the interference with value entry in the number input. Because `RangeControl` sends out clamped values if out-of-range values are entered in the number input, it has to avoid updating the number input with the clamped value coming back in from props. It's tricky and I couldn't find a simpler way to go about it than to imperatively overwrite the update to temporarily keep the divergence in the DOM ff6e10a.

38c0158, 8930905 & cb00749
These relate to fixing the reset behavior. 38c0158 fixes the reset to `initialPosition` but turned out to do so only for the uncontrolled instances. I noticed the reset behavior for controlled instances was busted again but the unit tests would not fail unless a `value` was defined 8930905. The tricky part about resets is that `undefined` may be sent out through `onChange` and, if `undefined` comes back in from props, the component has to distinguish it as a reset as opposed to a uncontrolled context. It was going to require both an extra state and effect hook to do that while employing `useControlledValue` and I discovered that `useControlledState` wouldn't need any extra help so I switched to the latter cb00749. I'd be glad to elaborate on this if anyone is interested in more detail.

fa2723d
While exploring the edge case of non-numeric values being entered in the number input I found that the state syncing in `InputControl` could get stuck due to one of the layout effect hooks not running when focus changes. I included the focus state variable in the hooks dependencies.


##### Before fa2723d 
https://user-images.githubusercontent.com/9000376/166313658-5194813b-e752-442f-9cf6-87629b314771.mp4
##### After
https://user-images.githubusercontent.com/9000376/166313786-822e4730-266d-4e87-881b-c1e3183c124f.mp4

**Comment by ciampo** [comment]
> A patch applicable to trunk with the same changes to range-control’s unit tests to verify that they all pass.

I can confirm that those tests on `RangeControl` all pass on `trunk`. Curiously enough, I also applied [that small change](https://github.com/WordPress/gutenberg/pull/40568/files#diff-776f85d0e9f680c2a229b53c23ceff84db7c2c221c5e1311ad3f1c14d8ba99a3R218) to `UnitControl`'s unit tests on `trunk`, and somehow that is causing a failure:

Screenshot 2022-05-02 at 21 36 36

> So the remaining two `RangeControl` issues were resets to `initialPosition` and potential interference with value entry in the number input. I think everything is finally solved here. Here's bit more info on the latest commits.

This is great to hear! It's quite late for me to have a look today, but I'll definitely do so tomorrow. Thank you once again for your efforts here 🙏

**Comment by stokesman** [comment]
>Curiously enough, I also applied [that small change](https://github.com/WordPress/gutenberg/pull/40568/files#diff-776f85d0e9f680c2a229b53c23ceff84db7c2c221c5e1311ad3f1c14d8ba99a3R218) to UnitControl's unit tests on trunk, and somehow that is causing a failure:

Yes, it seems on trunk (and here) that suite is just somehow working and any little change is likely to make some other test fail. Notice that the test that failed was not the one with the change. I think it's because there are actually a few more places in that file with a missing `await`. Here’s one:
https://github.com/WordPress/gutenberg/blob/ac5dd04c289977baf2a2fe37e8a9ed10f4e67b36/packages/components/src/unit-control/test/index.js#L542
We could fix the others in this PR or if we did it in a separate PR I don't think it would even cause a hiccup here.

File: packages/components/src/unit-control/test/index.js [hunk 2]
```diff
@@ -215,7 +215,7 @@ describe( 'UnitControl', () => {
 			expect( input.value ).toBe( '300px' );
 			expect( state ).toBe( 50 );
 
-			user.keyboard( '{Escape}' );
```
Comment (ciampo) [inline packages/components/src/unit-control/test/index.js:218]: This change can be dropped after rebasing on top of `trunk` since it was included in #40790

File: packages/components/src/unit-control/index.tsx [hunk 3]
```diff
@@ -178,10 +178,7 @@ function UnforwardedUnitControl(
 				: undefined;
 			const changeProps = { event, data };
 
-			onChangeProp?.(
-				`${ validParsedQuantity ?? '' }${ validParsedUnit }`,
-				changeProps
-			);
```
Comment (ciampo) [inline packages/components/src/unit-control/index.tsx:184]: These changes have been extracted to #40796, and can therefore be removed from this PR once #40796 is merged (and this PR is rebased on top of `trunk`).

File: packages/components/src/range-control/utils.js [hunk 4]
```diff
@@ -2,69 +2,48 @@
...
+	};
+	useEffect( () => {
+		if ( ref.current && isDiverging.current ) {
+			const input = ref.current;
+			const entry = input.value;
+			const { defaultView } = input.ownerDocument;
+			defaultView.requestAnimationFrame( () => ( input.value = entry ) );
+			isDiverging.current = false;
+		}
+	}, [ value ] );
```
Comment (ciampo) [inline packages/components/src/range-control/utils.js:49]: It would be good to add some inline comments here — eg. why are we saving the input's value, and reapplying it to the same input in the next rendering frame?

**Review by ciampo — APPROVED** [review]
Thank for for the hard work and the thoughtful explanations, @stokesman !

Honestly, the complexity that this component has reached makes it very complicated to code review. I did my best, but there's the chance I may have still missed something. At least the core behavior should be verified by the unit tests that we have in place.

Once the [last pending comment](https://github.com/WordPress/gutenberg/pull/40568#discussion_r865804080), feel free to merge this PR into #40518.

At that point, we will be able to:
- rebase on top of `trunk` and remove the changes that have already been merged in separate PRs ([one](https://github.com/WordPress/gutenberg/pull/40568#discussion_r864008809) and [two](https://github.com/WordPress/gutenberg/pull/40568#discussion_r864040386))
- Give a final, through round of review to #40518 before hopefully being able to merge this fix

(cc @youknowriad )

File: packages/components/src/input-control/reducer/reducer.ts [hunk 5]
```diff
@@ -206,31 +199,33 @@ export function useInputControlStateReducer(
...
 			onChangeHandler( state.value ?? '', {
-				event: currentState.current._event as
+				event: refEvent.current as
 					| ChangeEvent< HTMLInputElement >
 					| PointerEvent< HTMLInputElement >,
 			} );
 		}
-	}, [ state.value ] );
+		refEvent.current = null;
+	}, [ state.value, state.isDragging, isFocused ] );
```
Comment (youknowriad) [inline packages/components/src/input-control/reducer/reducer.ts]: It's not clear to me why we need to pass "isFocused" to this hook. For me, isFocused is more a "component" problem in the sense that if we want to trigger onChangeHandler when the component loses focus, we should just call `commit` (which will change the state.value and trigger the change handler. 

Adding isFocused and state.isDragging here is a code smell to me and potentially something that will create hidden bugs.

Comment (stokesman) [reply packages/components/src/input-control/reducer/reducer.ts]: The additional dependencies were added only in the interest of executing `refEvent.current = null` not `onChangeHandler`. That's done as part of making sure the effect hook used to intake values doesn't dispatch (via `reset` in the base PR) more than necessary.

Without using the existence of the event to avoid some dispatches, there was [an issue](https://github.com/WordPress/gutenberg/runs/6166230703?check_suite_focus=true#step:5:920) with `FocalPointPicker`. That component is surely doing it wrong but it seemed the shorter and more defensive path to reduce the dispatches.

I'm glad to take another look and see how it might be improved upon. I have some ideas.

Comment (stokesman) [reply packages/components/src/input-control/reducer/reducer.ts]: I've pushed 2b14b0f, which removes the additional dependencies. The event ref (if it exists) is now cleared when the intake effect hook runs. This is much better. Thank you for the critique @youknowriad 🙇

File: packages/components/src/range-control/utils.js (hunk 4 above)
Comment (stokesman) [reply packages/components/src/range-control/utils.js:49]: I added a comment in 085fd03 and then while double checking things while testing out the change that was to be 2b14b0f, I discovered the hook didn't need to handle `NaN` values so I made 29634d7. The issue (as non-critical as could be) was the first update after a `NaN` entry failing to re-sync the input:

https://user-images.githubusercontent.com/9000376/167703859-42878b30-ec8d-4dd1-907d-ea92d288f3fa.mp4

I'd missed this before https://github.com/WordPress/gutenberg/pull/40568#issuecomment-1115271440 because I dragged the slider instead of clicking it once.

**Comment by ciampo** [comment]
Just popping by to say that I won't be around until Wednesday (18th), so feel free to merge this PR into #40518 whenever Riad agrees to. We can then all take that PR for a solid, deep test before merging into `trunk`

File: packages/components/src/input-control/reducer/reducer.ts [hunk 6]
```diff
@@ -52,6 +52,14 @@ function inputControlStateReducer(
 	composedStateReducers: StateReducer
 ): StateReducer {
 	return ( state, action ) => {
+		// Updates state and returns early when there's no action type. These
+		// are controlled updates and need no exposure to additional reducers.
```
Comment (youknowriad) [inline packages/components/src/input-control/reducer/reducer.ts:56]: When does this happen (action without type)?

**Review by youknowriad — COMMENTED** [review]
To be completely honest, I don't really understand all the changes here but I'm ok with merging both PRs to test properly.

File: packages/components/src/input-control/reducer/reducer.ts (hunk 6 above)
Comment (stokesman) [reply packages/components/src/input-control/reducer/reducer.ts:56]: That would be in what I'm calling the intake effect https://github.com/WordPress/gutenberg/blob/2b14b0fa9cf8a253a884195b9d8bb419542a43f6/packages/components/src/input-control/reducer/reducer.ts#L218-L226

That was dispatching via `reset` in the base PR but in looking for a way to distinguish the update from props I landed on making the bare `dispatch` with no action type. I'd considered implementing another action (or bringing back `update` that used to exist and was used only for this purpose (props coming in)) but I went with the bare `dispatch` because I thought it would make for a little less changes.

Comment (youknowriad) [reply packages/components/src/input-control/reducer/reducer.ts:56]: I'd have just used another action but it's not important :)

**Comment by stokesman** [comment]
Ideally there might have been less roundabout on the way to these changes. I'd much prefer that the review burden were less but I'm not doubtful of the result. Thanks again to all involved for the help!
